lib/portus/background/sync.rb

```diff
@@ -0,0 +1,70 @@
+# frozen_string_literal: true
+
+require "portus/errors"
+
+module Portus
+  module Background
+    # Sync synchronizes the contents from the registry into the DB.
+    class Sync
+      # Returns how many seconds has to pass between each loop for this
+      # synchronization to happen.
+      def sleep_value
+        20
+      end
+
+      # Returns always true because this does not depend on some configuration
+      # options.
+      def work?
+        true
+      end
+
+      # Simply fetches the catalog from the registry and calls
+      # `update_registry!`.
+      def execute!
+        ::Registry.find_each do |registry|
+          begin
+            cat = registry.client.catalog
+            Rails.logger.debug "Catalog:\n #{cat}"
+
+            # Update the registry in a transaction, since we don't want to leave
+            # the DB in an unknown state because of an update failure.
+            ActiveRecord::Base.transaction { update_registry!(cat) }
+          rescue *::Portus::Errors::NET,
+                 ::Portus::Errors::NoBearerRealmException, ::Portus::Errors::AuthorizationError,
+                 ::Portus::Errors::NotFoundError, ::Portus::Errors::CredentialsMissingError => e
```

Comment:
Is it really necessary to catch all there exceptions? Does all of them make sense? What was wrong with the `StandardError`?

Reply:
The problem is that `StandardError` is the base for all exceptions, and you should never rescue that, because you'd be rescuing legit bugs too. This is why I had to re-write some tests: the registry hostname was wrong in tests (see the registry factory), and so an exception was raised about being an invalid URL, but it was catched as a StandardError. Now, that is not something we want (and plus it makes debugging harder...).

So, instead, what we do now is:

- Let's rescue network-related exceptions.
- Let's rescue registry-related exceptions since we are interfacing the registry here. I wanted to group these exceptions too, but I was getting a weird error when doing so, and after some attempts I went back to listing all registry errors and be done with it 😅

Reply:
Also notice that the old code had a `TODO` about this as well 😀

Reply:
It makes sense to catch only the interested exceptions. The reason I said that was because some of those would never happen in this case (e.g. `NoBearerRealmException`). I'm totally fine with catching specific exceptions, I just thought that it was too much and unnecessary.

Reply:
This may happen during the authentication process. It's maybe a bit too much, but I don't know, it's better to be safe than sorry I guess 😅

Reply:
> This may happen during the authentication process.

This is running in the background and calling the db directly, how is authentication being applied to this? I know this doesn't affect the code but what I wanted to avoid is the unnecessary pollution putting all these exceptions everywhere.

Reply:
This background service fetches the catalog. In order to do so it logs in as the portus user (in the registry API) and performs an API call.

Reply:
Ok, I didn't realize the details. 😅

Reply:
No worries 😉